.github/workflows/continuous-integration-workflow.yml

```diff
@@ -7,12 +7,42 @@ on:
       - 'README.md'
 
 jobs:
-  build:
+  build-and-test:
     name: Build and test
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
     - name: build docker container 
       run: docker build --target test --tag todo-app:test .
     - name: run docker tests
-      run: docker run todo-app:test
\ No newline at end of file
+      run: docker run todo-app:test
+      
+  build-and-push-docker-image:
+    name: Docker Build, Tag & Publish
+    needs: build-and-test
+    if: ${{ github.ref == 'refs/heads/main' }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout Code
+        uses: actions/checkout@v1
+        id: git_checkout
+
+      - name: Build, Tag & Publish Image to DockerHub
+        uses: dinushchathurya/build-tag-push-action@v1.2.1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+          context: .
+          file: Dockerfile
+          repoOwner: ${{ secrets.DOCKERHUB_USERNAME }}
+          repository:  todo-app
+          tag: ${{github.ref_name}}
```

Comment:
Given that you only want to end up running this stage on your main branch - would hardcoding this make more sense? I can imagine that changing our default/prod branch wouldn't necessarily mean we want to change our Docker image's tag